linked_list/linked_list.py

```diff
@@ -9,71 +9,140 @@ def __init__(self, value, next_node = None):
 # Defines the singly linked list
 class LinkedList:
     def __init__(self):
-      self.head = None # keep the head private. Not accessible outside this class
+        self.head = None # keep the head private. Not accessible outside this class
 
     # returns the value in the first node
     # returns None if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def get_first(self):
-        pass
+        if not self.head:
+            return None
+        return self.head.value
 
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(self, value):
-        pass
+        self.head = Node(value, self.head)
+
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def search(self, value):
-        pass
+        if not self.head:
+            return False
+        current = self.head
+        while current:
+            if current.value == value: 
+                return True
+            current = current.next
+        return False
+            
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):
-        pass
+        if not self.head:
+            return 0
+        length = 0
+        current = self.head
+        while current:
+            current = current.next
+            length += 1
+        return length
+
+
+
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity:O(1)
     def get_at_index(self, index):
-        pass
+        if not self.head:
+            return
+        current = self.head
+        current_index = 0
+        while current:
+            if current_index == index:
+                return current.value
+            current = current.next
+            current_index +=1
+            
+        
+        
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: 0(1)
     def get_last(self):
-        pass
+        if not self.head:
+            return
+        current = self.head
+        while current:

```

Comment:
You could also do a `while current.next:` and then you don't need the following if statement.